owncloudApp/src/main/java/com/owncloud/android/ui/activity/PassCodeActivity.kt

```diff
@@ -0,0 +1,425 @@
+                finish()
+            } else {
+                showErrorAndRestart(
+                    R.string.pass_code_wrong, R.string.pass_code_enter_pass_code,
+                    View.INVISIBLE
+                )
+            }
+        } else if (ACTION_REQUEST_WITH_RESULT == intent.action) {
+            /// enabling pass code
+            if (!mConfirmingPassCode) {
+                mPassCodeError.visibility = View.INVISIBLE
+                requestPassCodeConfirmation()
+            } else if (confirmPassCode()) {
+                /// confirmed: user typed the same pass code twice
+                savePassCodeAndExit()
+            } else {
+                showErrorAndRestart(
+                    R.string.pass_code_mismatch, R.string.pass_code_configure_your_pass_code, View.VISIBLE
+                )
+            }
+        }
+    }
+
+    private fun showErrorAndRestart(
+        errorMessage: Int, headerMessage: Int,
+        explanationVisibility: Int
+    ) {
+        Arrays.fill(mPassCodeDigits, null)
+        mPassCodeError.setText(errorMessage)
+        mPassCodeError.visibility = View.VISIBLE
+        mPassCodeHdr.setText(headerMessage) // TODO check if really needed
+        mPassCodeHdrExplanation.visibility = explanationVisibility // TODO check if really needed
+        clearBoxes()
+    }
+
+    /**
+     * Ask to the user for retyping the pass code just entered before saving it as the current pass
+     * code.
+     */
+    protected fun requestPassCodeConfirmation() {
+        clearBoxes()
+        mPassCodeHdr.setText(R.string.pass_code_reenter_your_pass_code)
+        mPassCodeHdrExplanation.visibility = View.INVISIBLE
+        mConfirmingPassCode = true
+    }
+
+    /**
+     * Compares pass code entered by the user with the value currently saved in the app.
+     *
+     * @return     'True' if entered pass code equals to the saved one.
+     */
+    protected fun checkPassCodeIsValid(): Boolean {
+        val appPrefs = PreferenceManager
+            .getDefaultSharedPreferences(applicationContext)
+        val passcodeString = appPrefs.getString(PREFERENCE_PASSCODE, loadPinFromOldFormatIfPossible())
+        var isValid = true
+        var i = 0
+        while (i < mPassCodeDigits.size && isValid) {
+            val originalDigit = Character.toString(passcodeString!![i])
+            isValid = mPassCodeDigits[i] != null && mPassCodeDigits[i] == originalDigit
+            i++
+        }
+        return isValid
+    }
+
+    /**
+     * Compares pass code retyped by the user in the input fields with the value entered just
+     * before.
+     *
+     * @return     'True' if retyped pass code equals to the entered before.
+     */
+    protected fun confirmPassCode(): Boolean {
+        mConfirmingPassCode = false
+        var isValid = true
+        var i = 0
+        while (i < mPassCodeEditTexts.size && isValid) {
+            isValid = mPassCodeEditTexts[i]!!.text.toString() == mPassCodeDigits[i]
```

Comment:
There are a lot of `!!` in this file, we should try to move to a null safe syntax.